core/auth_provider_okta.go

```diff
@@ -0,0 +1,311 @@
+	AuthorizationServer string `json:"authorization_server"`
+	OktaDomain		 	string `json:"okta_domain"`
+	DeploymentURI		string `json:"deployment_uri"`
+
+	OktaEnterprise 	 bool   `json:"okta_enterprise"`
+	Mapping          []struct {
+		Okta string `json:"okta"`
+		Tenant string `json:"tenant"`
+		Rights []struct {
+			Group string `json:"group"`
+			Role string `json:"role"`
+		} `json:"rights"`
+	} `json:"mapping"`
+}
+
+type Exchange struct {
+	Error            string `json:"error,omitempty"`
+	ErrorDescription string `json:"error_description,omitempty"`
+	AccessToken      string `json:"access_token,omitempty"`
+	TokenType        string `json:"token_type,omitempty"`
+	ExpiresIn        int    `json:"expires_in,omitempty"`
+	Scope            string `json:"scope,omitempty"`
+	IdToken          string `json:"id_token,omitempty"`
+}
+
+func (p *OktaAuthProvider) Configure(raw map[interface{}]interface{}) error {
+	b, err := json.Marshal(util.StringifyKeys(raw))
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(b, p)
+	if err != nil {
+		return err
+	}
+
+	if p.ClientID == "" {
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `client_id' value")
+	}
+
+	if p.ClientSecret == "" {
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `client_secret' value")
+	}
+
+	if p.OktaDomain == "" {
+		// ex: p.OktaDomain = "https://dev-xyz.okta.com/" 
+		// p.OktaIssuer = "https://dev-xyz.okta.com/oauth2/default" 
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `okta_domain' value")
+	}
+
+	if p.DeploymentURI == "" {
+		// ex: p.DeploymentURI = "https://shield.starkandwayne.com" 
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `deployment_uri' value")
+	}
+
+	if p.AuthorizationServer == "" {
+		// ex: p.AuthorizationServer = "default" 
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `authorization_server' value")
+	}
+
+
+	p.OktaDomain = strings.TrimSuffix(p.OktaDomain, "/")
+	p.DeploymentURI = strings.TrimSuffix(p.DeploymentURI, "/")
+
+
+	p.properties = util.StringifyKeys(raw).(map[string]interface{})
+
+	return nil
+}
+
+func (p *OktaAuthProvider) WireUpTo(c *Core) {
+	p.core = c
+}
+
+func (p *OktaAuthProvider) ReferencedTenants() []string {
+	ll := make([]string, 0)
+	for _, m := range p.Mapping {
+		ll = append(ll, m.Tenant)
+	}
+	return ll
+}
+
+func (p *OktaAuthProvider) Initiate(r *route.Request) {
+
+	var state = randstr.Hex(16)
+
+	var redirectPath string
+
+	q := r.Req.URL.Query()
+	q.Add("client_id", p.ClientID)
+	q.Add("response_type", "code")
+	q.Add("response_mode", "query")
+	q.Add("scope", "openid profile email groups")
+	q.Add("redirect_uri", p.DeploymentURI + "/auth/okta/redir") //ex: https://shield.starkandwayne.com/auth/okta/redir
+	q.Add("state", state)
+	// q.Add("nonce", nonce)
+
+	redirectPath = p.OktaDomain + "/oauth2/" + p.AuthorizationServer + "/v1/authorize?" + q.Encode()
```

Comment:
For completeness and cleanliness, whenever dealing with URLs, consider using the URL{} struct from the url standard package. https://pkg.go.dev/net/url#URL

Reply:
type URL struct {
	Scheme      [string](https://pkg.go.dev/builtin#string)
	Opaque      [string](https://pkg.go.dev/builtin#string)    // encoded opaque data
	User        *[Userinfo](https://pkg.go.dev/net/url#Userinfo) // username and password information
	Host        [string](https://pkg.go.dev/builtin#string)    // host or host:port
	Path        [string](https://pkg.go.dev/builtin#string)    // path (relative paths may omit leading slash)
	RawPath     [string](https://pkg.go.dev/builtin#string)    // encoded path hint (see EscapedPath method)
	OmitHost    [bool](https://pkg.go.dev/builtin#bool)      // do not emit empty host (authority)
	ForceQuery  [bool](https://pkg.go.dev/builtin#bool)      // append a query ('?') even if RawQuery is empty
	RawQuery    [string](https://pkg.go.dev/builtin#string)    // encoded query values, without '?'
	Fragment    [string](https://pkg.go.dev/builtin#string)    // fragment for references, without '#'
	RawFragment [string](https://pkg.go.dev/builtin#string)    // encoded fragment hint (see EscapedFragment method)
}

It does not coincide with the values we have.

Reply:
It would be like the following:
```golang
u := url.URL{
  Scheme: strings.Split(p.OktaDomain, "://")[0],
  Host: strings.Split(p.OktaDomain, "://")[1],
  Path: url.JoinPath("/oauth2/", p.AuthorizationServer, "/v1/authorize"),
  RawQuery: q.Encode(),
}
```

Maybe consider splitting scheme and host in the `OktaAuthProvider` definition.